<span style="background-color:#eee">Input files for MIMAR software </span>
---------------------
```
Name length x_y v_y w_y n_1 n_2 S_1 S_2 S_s S_f //
```
> With
>
> - length: number of synonymous sites (obtained with DnaSP)
>
> - x_y: inheritance scalar (1: autosomal, 0.75: X, 0.25: Y/mtDNA)
>
> - v_y: mutation rate variation scalar = D(JC)/Average(DC(JC))  
>   (DC(JC) = Divergence Jukes Cantor obtained with DnaSP) 
>
> - w_y: recombination inheritance and rate variation scalar: ratio of the locus-specific population recombination rate per base over rho (0: mtDNA/Y, 1(0.5): autosomes in Mammals (*Drosophila*))
>
> - n_*i*: sample size
>
> - S_*i*: summary of polymorphism statistics 

### How to obtain polymorphism statistics:

#### Input files
- "synsites.in": position of all synonymous + non-coding sites (obtained with DnaSP) in 1 column
- "exon.in": 
<ul> <li> line1 = sequence_length  number_of_seq_in_pop1 number_of_seq_in_pop2 
<li> other lines: sequences (1st seq = outgroup)</li></ul>

#### Script
```perl
#!/usr/bin/perl 
use strict;


my $inputfile = $ARGV[0];
my $inputfile2 = $ARGV[1];
my $outputfile = $ARGV[2];

#### Prepare the input file with sequences

unless (open DATAINPUT, $inputfile) {
	print "Can not find the data input file!\n";
	exit;
}

#my $outputfile = "4Ss_y.out";

unless (open(OUTFILE, ">$outputfile"))  {
	print "I can\'t write to $outputfile   $!\n\n";
	exit;
}
print "Creating output file: $outputfile\n";



###################
my $linenumber=0;
my %hash;
my @numbers;
my @temp;
my @temp1;
my $m;
my $n;
my $r;
my $invariant=0;
my $invariantpop1=0;
my $invariantpop2=0;
my $ancestralpop1;
my $ancestralpop2;
my @S1;
my @S2;
my @S3;
my @S4;
my @PD;

my @synsites;

my @SynS1;
my @SynS2;
my @SynS3;
my @SynS4;
my @SynPD;


while ( my $line = <DATAINPUT>) {
## set first line equal to gene name

	if ($linenumber == 0)  { # this is the first line
		@numbers = split(' ',$line);
		print  "The number of bases is: $numbers[0] \n";
		print  "The number of seqs in pop1 is $numbers[1] \n";
		print  "The number of seqs in pop2 is $numbers[2] \n";
				$linenumber += 1;
	}
	elsif (($linenumber ne 0) && ($linenumber <= ($numbers[1]+$numbers[2]+1))) { 
	# this is the line containing the sequences (the first is the outgroup)
		@temp = split(' ',$line);
		$hash{$linenumber-1}[0]=$temp[0];
		@temp1=split('',$temp[1]);
		for ($m=0; $m <= $#temp1; $m++){
			$hash{$linenumber-1}[$m+1]=$temp1[$m];
		}
		@temp=();  #this should erase the contents of the array
		@temp1=();  #this should erase the contents of the array
				$linenumber += 1;
	} 
}	#end while


#### Prepare the input file with synonymous sites

unless (open DATAINPUT2, $inputfile2) {
	print "Can not find the synonymous sites input file!\n";
	exit;
}
###################

while ( my $line = <DATAINPUT2>) {
		@temp = split(' ',$line);
		push(@synsites,$temp[0]);
}	#end while



# now the has should be loaded with a name in [0] and bp in all of the rest.

#cycle through each position and categorize it
for ($n=1; $n <= $numbers[0]; $n++){

		#first test if the site is invariant in ingroup
		for ($r=2; $r <= ($numbers[1]+$numbers[2]); $r++){
				if (uc $hash{1}[$n] ne uc $hash{$r}[$n]){
						$invariant+=1;
				}				
		} #end $r
		if ($invariant ne 0){ #the site was not invariant in the ingroup
			#test whether the site is constant in pop1, and whether pop1 has ancestral state
			for ($r=2; $r <= ($numbers[1]); $r++){
				if (uc $hash{1}[$n] ne uc $hash{$r}[$n]){
						$invariantpop1+=1;
				}
			}		
			for ($r=1; $r <= ($numbers[1]); $r++){
				if (uc $hash{0}[$n] eq uc $hash{$r}[$n]){
						$ancestralpop1+=1;
				}		
			}
			
			#test whether the site is constant in pop2
			for ($r=($numbers[1]+2); $r <= ($numbers[1]+$numbers[2]); $r++){
				if (uc $hash{$numbers[1]+1}[$n] ne uc $hash{$r}[$n]){
						$invariantpop2+=1;
				}
			}
			for ($r=($numbers[1]+1); $r <= ($numbers[1]+$numbers[2]); $r++){
				if (uc $hash{0}[$n] eq uc $hash{$r}[$n]){
						$ancestralpop2+=1;
				}		
			}

		}#endif
	
	
	#Now decide what kind of site it is
	#if it is invariant do not do anything
		if ($invariant ne 0){	
			if (($invariantpop1 eq 0) && ($invariantpop2 eq 0) && ($ancestralpop1 ne 0) && ($ancestralpop2 eq 0)) {
			# pop2 must be fixed for a derived allele
					push(@S4,$n); #add the character position to @S4
			}
			elsif (($invariantpop1 eq 0) && ($invariantpop2 eq 0) && ($ancestralpop1 eq 0) && ($ancestralpop2 ne 0)){
			# pop1 must be fixed for a derived allele
					push(@S4,$n); #add the character position to @S4
			}
			elsif (($invariantpop1 ne 0) && ($invariantpop2 eq 0) && ($ancestralpop1 ne 0) && ($ancestralpop2 ne 0)) {
			# pop1 has a derived and an ancestral allele; pop2 fixed for ancestral
					push(@S1,$n); #add the character position to @S1
			}
			elsif(($invariantpop1 eq 0) && ($invariantpop2 ne 0) && ($ancestralpop1 ne 0) && ($ancestralpop2 ne 0)) {
			# pop2 has a derived and an ancestral allele; pop2 fixed for ancestral
					push(@S2,$n); #add the character position to @S2
			}
			elsif(($invariantpop1 ne 0) && ($invariantpop2 ne 0) && ($ancestralpop1 ne 0) && ($ancestralpop2 ne 0)) {
			# pop1 and pop2 have derived and ancestral alleles
					push(@S3,$n); #add the character position to @S3
			}
			elsif(($invariantpop1 ne 0) && ($invariantpop2 eq 0) && ($ancestralpop1 ne 0) && ($ancestralpop2 eq 0)) {
			# pop1 and pop2 have derived and ancestral alleles
					push(@S3,$n); #add the character position to @S3
			}
			elsif(($invariantpop1 eq 0) && ($invariantpop2 ne 0) && ($ancestralpop1 eq 0) && ($ancestralpop2 ne 0)) {
			# pop1 and pop2 have derived and ancestral alleles
					push(@S3,$n); #add the character position to @S3
			}
			elsif(($invariantpop1 ne 0) && ($invariantpop2 ne 0) && ($ancestralpop1 eq 0) && ($ancestralpop2 eq 0)){
			# pop1 and pop2 both are variable but all are diverged from outgroup
					push(@PD,$n); #add the character position to @PD
			}
			elsif(($invariantpop1 eq 0) && ($invariantpop2 ne 0) && ($ancestralpop1 eq 0) && ($ancestralpop2 eq 0)){
			# pop2 is variable, pop1 isnt but both have only derived alleles
					push(@PD,$n); #add the character position to @PD
			}
			elsif(($invariantpop1 ne 0) && ($invariantpop2 eq 0) && ($ancestralpop1 eq 0) && ($ancestralpop2 eq 0)){
			# pop1 is variable, pop2 isnt but both have only derived alleles
					push (@PD,$n); #add the character position to @PD
			}	
			else{
					push (@PD,$n); #add the character position to @PD	
			}
		} #endif
	
	
	#reset the variables
	$invariant=0;
	$invariantpop1=0;
	$invariantpop2=0;
	$ancestralpop1=0;
	$ancestralpop2=0;	
	} # end $n



# now make new arrays that contain only synonymous sites

		for ($r=0; $r <= $#S1; $r++){
			for ($n=0; $n <= $#synsites; $n++){	
				if($S1[$r] eq $synsites[$n]){ #add it to the syn array
					push (@SynS1,$S1[$r]);
				}		
			}
		}
		for ($r=0; $r <= $#S2; $r++){
			for ($n=0; $n <= $#synsites; $n++){	
				if($S2[$r] eq $synsites[$n]){ #add it to the syn array
					push (@SynS2,$S2[$r]);
				}		
			}
		}
		for ($r=0; $r <= $#S3; $r++){
			for ($n=0; $n <= $#synsites; $n++){	
				if($S3[$r] eq $synsites[$n]){ #add it to the syn array
					push (@SynS3,$S3[$r]);
				}		
			}
		}
		for ($r=0; $r <= $#S4; $r++){
			for ($n=0; $n <= $#synsites; $n++){	
				if($S4[$r] eq $synsites[$n]){ #add it to the syn array
					push (@SynS4,$S4[$r]);
				}		
			}
		}

print "Position of sites only segregating in population 1: ", "@SynS1","\n";
print "Position of sites only segregating in population 2: ", "@SynS2","\n";
print "Position of sites only segregating in pop1 and pop2: ", "@SynS3","\n";
print "Position of derived sites fixed on pop1 or pop2: ", "@SynS4","\n";
print "Position of other sites beyond characterization because outgroup is completely diverged from ingroup: ", "@PD","\n";

print OUTFILE "SynS1 ",, "@SynS1","\n";
print OUTFILE "SynS2 ",, "@SynS2","\n";
print OUTFILE "SynS3 ",, "@SynS3","\n";
print OUTFILE "SynS4 ",, "@SynS4","\n";


print "Number of synonymous S1 sites: ", ($#SynS1+1),"\n";
print "Number of synonymous S2 sites: ", ($#SynS2+1),"\n";
print "Number of synonymous S3 sites: ", ($#SynS3+1),"\n";
print "Number of synonymous S4 sites: ", ($#SynS4+1),"\n";

print "You  still need to check to see if there are more than 2 SNPs per segregating site!!!\n";
						
#  close the output file
close OUTFILE;   
print "Closing output file: $outputfile\n";

#   close DATAINPUT;
close DATAINPUT;

exit;
```
### To run the script

```
Directory_name location $ perl script_name exon.in synsites.in output_file_name.out

```

### Output files

The program calculates 3 types of derived segregating sites:
<ul> <li> those specific to population 1
<li> those specific to population 2
<li> those shared by population 1 and 2
<li> those fixed in population 1 or population 2 </p></li>
</ul>

In the output file, we have the position of each sites corresponding to the different types of segregating sites (line1: S1=specific to population 1, S2=specific to population 2, S_s=shared by population 1 and 2, S_f=fixed in population 1 or population 2)

<span style="background-color:#eee">How to specify a model (on kraken on [Sharcnet](https://www.sharcnet.ca/my/front/)) </span> 
---------------------
```
sqsub -r duration -o outputmimar ./mimar nsteps bsteps Y -lf input -u nu -t teta1 -n teta2 -ej T -N tetaA -M M -i int -x ngen -yt -L osteps -v V(teta1) V(teta2) V(T) V(tetaA) V(M12) V(M21) -r rho -o exsoutput

```
> With:
>
>  - duration: time allowed on sharcnet to run the job
>  nsteps: total number of steps (if "-yt" is useds, in minutes) = burnin + following MCMC steps
>
>  - bsteps: number of burnin
>
>  - Y: number of *loci*
>
>  - nu: generational mutation rate / bp
>
>  - teta1, teta2, tetaA: population mutation rates / bp. If teta2 & tetaA not specified: equal to teta1. If we use "u a b" as a teta, it means Unif[a,b]
>
>  - T: split time in generaions
>
>  - int: interval between accepted parameters
>
>  - ngen: number of genealogies to estimate the likelihood at each *locus*
>
>  - osteps: number of steps separating intermediate summary outputfiles ("-yt": in minutes)
>
>      <font color='red'>! Be careful</font>: if osteps < bsteps: empty histograms 
>
>  - V(teta*i*): variances for the kernel distribution of parameters
>
>  - rho: population cross-over rate parameter
